capture.go

```diff
@@ -68,63 +82,188 @@ func WithDir(dir string) Option {
 }
 
 // New creates a Capture struct with (optionally) Env and Dir set on it.
-func New(options ...Option) Capture {
+func New(options ...Option) *Capture {
 	var c Capture
 	for _, option := range options {
 		option(&c)
 	}
 
-	return c
+	return &c
 }
 
-// Run executes a name and returns Capture and error.
-//
-// The Capture will always be returned, even if it's incomplete.
+// Run executes a name with args and returns an error.
 //
 // The error type is passed on unwrapped and may contain an *exec.ExitError, which can be converted with errors.As
 // to check for additional information.
 //
 // The parameters behave as they do in *exec.Cmd, so passing in a nil env will inherit the parent environment,
 // and passing an empty slice will create and empty environment.
-func (c Capture) Run(ctx context.Context, name string, args ...string) (Capture, error) {
-	if ctx == nil {
```

Comment:
This is more of a curiosity question as dealing with "developer" error is something I struggle with. While it seems sane to check for whether or not ctx may have been nil, is it worth building those safeguards into the application just because it *could* be? Most of the time I think I follow this same pattern of adding the check and then pulling it away as the shape of the interaction starts to solidify

Reply:
Context is a really weird one. It didn't exist until the creation of `/x/net/context` in 2014. So from 2008-2014 it didn't exist and a lot of functions got on fine without it. (they'd use some other signaling)

It didn't enjoy its fully implemented status until ca. 2016 when it was moved into the main `/pkg/` stable. The advice has always been to use the context version of a thing, to put `context.Context` (or `ctx`) as the first parameters of context-aware functions, just as we commonly use `err error` as the second return parameter of functions that **may** fail.

While this could be considered a developer error, the reason I replace a `nil` with `context.Backgroud()` is to just correct for the occasional `nil` without ill effects.

Now, I could do the same pre-to-post "context-world" thing and have two versions of these functions, with the short declaration wrapping the `context.Background()` call and the long version just passing on the `nil` context. But I figured a little anti-fragility here goes a long way to prevent random panics, and that's a good thing.

Whether we should always check incoming types for their existential or boundary values depends very much on the function, but I think in *this case* amongst case-by-case basis is fine to prevent ID-10-T errors. 

There are a few other principles to keep in mind, like not calling `panic` from a library. The classic "be liberal in what you accept, strict in what you return" Robustness principle, and a few other things I'm not thinking about at the moment.